### ISSUE TITLE
make protocol more resilient to changes

### ISSUE THREAD

**Issue by rade** [weaveworks/weave#872]
At the moment every change to the protocol requires a change in version and introduces an incompatibility with all other versions.

We should make the protocol more resilient to change, permitting weave peers to communicate with each other when they speak slightly different protocol versions, without having to write any special code.

This works in conjunction with #871 by providing an easy route by to support multiple protocol versions.

Extracted from https://github.com/weaveworks/weave/issues/403#issuecomment-104048264.

**Comment by rade** [comment]
I've been contemplating whether we should have a major/minor protocol versioning scheme. The major version would be what gets agreed through #871. The minor version of that is then what the individual peers actually speak, and need not be the same for the two sides.

Without a minor version, we could, say, have Peer A supporting version 11 and 12, and Peer B supporting 12 and 13, both agreeing on 12, yet B actually speaking 13, because it knows that 13 can be handled by 12. This may look somewhat strange when examining the message stream, and generally makes it somewhat less obvious what is going on.

**Comment by awh** [comment]
It might be interesting to look back at all the changes that have precipitated protocol version bumps and see how they might have been handled under a major/minor scheme...

**Comment by rade** [comment]
Not very well at all. Not without extra code. I said that much when we had the discussion in email.

**Comment by bboreham** [comment]
I don't understand the usage of major/minor here. I would expect these terms to mean, in the example, that B supports 12.0 and 12.1, so 12 is the major version and .0 or .1 is the minor version.

**Comment by rade** [comment]
Yes, that is correct. Though I am not sure whether the minor version should really feature in the negotiation, since it should have no bearing on it. OTOH, there is no other place where the minor version would appear, which would make it pointless since it only exists for clarity.

**Comment by rade** [comment]
So after further discussion we have agreed that a single version number is sufficient. Let's keep things simple.

**Comment by rade** [comment]
It is already possible to [add new protocol messages](https://github.com/weaveworks/weave/blob/eae9928e98d14557ec1f0e59b63d7b7f1d11b294/router/connection.go#L478).

I have raised bugs for

- [ ] #911
- [ ] #913

The other area worth examining is the various gossip codecs, e.g. topology, ipam.
